File: apps/desktop/src/lib/sessions.ts
// Pure helpers for the session list (Issue #19). Kept free of React so the label
// resolution + filter are unit-testable and shared by the sidebar's display and
// its search — so what you type matches exactly what you see.

import type { Session } from "@/bindings";
import type { SearchHit } from "@/bindings/SearchHit";

/**
 * The label a user sees for a session: persisted title > goal > fallback.
 * `session.title` is server-truth (auto-derived from the first user message,
 * overridable via `rename_session`); the durable session store (ff-session,
 * RFC 0012) means the title survives a restart, so the frontend no longer
 * keeps a legacy localStorage fallback (#52).
 */
export function resolveLabel(session: Session): string {
  if (session.title) return session.title;
  if (session.goal) return session.goal;
  return "New session";
}

const FORK_SUFFIX_RE = / \(Fork (\d+)\)$/;

/** Strip a trailing " (Fork <k>)" suffix (#1069) so forking a fork renumbers
 *  from the original base title instead of stacking suffixes. */
export function stripForkSuffix(title: string): string {
  return title.replace(FORK_SUFFIX_RE, "");
}

function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

/**
 * Compute the next "<base> (Fork N)" title for a session forked from
 * `sourceTitle` (#1069). `base` strips any existing "(Fork k)" suffix from the
 * source so forking a fork stays on the same base and keeps numbering
 * contiguous; `N` is one past the highest existing "(Fork N)" sharing that
 * base among `existingTitles` (the in-memory session list), or 1 if none.
 */
export function nextForkTitle(
  sourceTitle: string,
  existingTitles: (string | null)[],
): string {
  const base = stripForkSuffix(sourceTitle);
  const re = new RegExp(`^${escapeRegExp(base)} \\(Fork (\\d+)\\)$`);
  let max = 0;
  for (const title of existingTitles) {
    if (!title) continue;
    const m = re.exec(title);
    if (m) max = Math.max(max, Number(m[1]));
  }
  return `${base} (Fork ${max + 1})`;
}

/**
 * Client-side filter: case-insensitive substring over the *resolved* label, so a
 * renamed title and a goal both match what the user actually sees. An empty or
 * whitespace-only query returns the list unchanged, order preserved.
 */
export function filterSessions(sessions: Session[], query: string): Session[] {
  const q = query.trim().toLowerCase();
  if (!q) return sessions;
  return sessions.filter((session) =>
    resolveLabel(session).toLowerCase().includes(q),
  );
}

/**
 * One content-search result to show as its own sidebar row (#710): the matched
 * session paired with its best hit (for the snippet + jump-to-message).
 */
export interface ContentHitRow {
  session: Session;
  hit: SearchHit;
}

/**
 * Reduce cross-session `searchMessages` hits (BM25 order, possibly several per
 * session) to one row per session for the sidebar's content group (#710):
 * - keep the first (best-ranked) hit per session, preserving BM25 order;
 * - drop sessions already shown as title matches (`excludeSessionIds`);
 * - drop hits whose session isn't in `sessionsById` (e.g. a not-yet-listed
 *   draft), so every row resolves to a real, visible session.
 */
export function groupContentHits(
  hits: SearchHit[],
  excludeSessionIds: ReadonlySet<string>,
  sessionsById: ReadonlyMap<string, Session>,
): ContentHitRow[] {
  const seen = new Set<string>();
  const rows: ContentHitRow[] = [];
  for (const hit of hits) {
    if (seen.has(hit.sessionId) || excludeSessionIds.has(hit.sessionId)) {
      continue;
    }
    const session = sessionsById.get(hit.sessionId);
    if (!session) continue;
    seen.add(hit.sessionId);
    rows.push({ session, hit });
  }
  return rows;
}

/**
 * Make an FTS5 `snippet()` string safe to inject as HTML (#710, PR #747 C1).
 * The backend wraps matched terms in `<mark>` but does NOT escape the
 * surrounding message text, which routinely contains raw HTML/JS (agents quote
 * `<script>`, `<img onerror=…>`, etc.). Rendering it verbatim in the Tauri
 * webview would be a live XSS with full IPC access. So we escape every text
 * segment and re-emit only the backend's own `<mark>`/`</mark>` delimiters.
 */
export function sanitizeSnippet(raw: string): string {
  return raw
    .split(/(<\/?mark>)/g)
    .map((part) =>
      part === "<mark>" || part === "</mark>" ? part : escapeHtml(part),
    )
    .join("");
}

function escapeHtml(s: string): string {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

/**
 * A short, human relative date for a search-hit row (#876): "Today"/"Yesterday"
 * for the last two calendar days (by local wall-clock day, not a rolling
 * 24h/48h window), else a short date ("Jul 3"), with the year appended only
 * when it differs from the current year ("Jul 3, 2024"). Distinct from
 * memory-view.ts's `humanAge()` ("~3 days ago" style), which the mockups don't
 * use. `Date.UTC` on each date's local Y/M/D is just an integer-day bucketing
 * trick to avoid DST arithmetic bugs from a naive ms-diff divide.
 */
export function formatHitDate(
  createdAtMs: number,
  now: number = Date.now(),
): string {
  const day = (ms: number) => {
    const d = new Date(ms);
    return Date.UTC(d.getFullYear(), d.getMonth(), d.getDate());
  };
  const diffDays = Math.round((day(now) - day(createdAtMs)) / 86_400_000);
  if (diffDays === 0) return "Today";
  if (diffDays === 1) return "Yesterday";
  const created = new Date(createdAtMs);
  const sameYear = created.getFullYear() === new Date(now).getFullYear();
  return created.toLocaleDateString("en-US", {
    month: "short",
    day: "numeric",
    year: sameYear ? undefined : "numeric",
  });
}

/** How many rows the list grows by per "Show more" click, and the initial
 *  batch size (#667). One endless, incremental reveal replaced the old
 *  All/Dismissed tabs + fixed cap. */
export const SESSION_REVEAL_BATCH = 25;

/**
 * Order the full session list for display (#667): pinned first, then the rest of
 * the non-dismissed sessions, then dismissed sessions last. Order within each of
 * the three groups is preserved (stable). Dismissed always sink to the bottom
 * regardless of pin, so a dismissed session never floats above a live one.
 */
export function arrangeSessions(
  sessions: Session[],
  pinned: ReadonlySet<string>,
  dismissed: ReadonlySet<string>,
): Session[] {
  const rank = (s: Session): number => {
    if (dismissed.has(s.id)) return 2; // dismissed: always last
    if (pinned.has(s.id)) return 0; // pinned live: first
    return 1; // other live
  };
  // Stable sort by group rank (Array.prototype.sort is stable), so insertion
  // order is kept within each group.
  return [...sessions].sort((a, b) => rank(a) - rank(b));
}

/**
 * Reveal the first `revealCount` of the arranged list, always keeping the active
 * session visible even if it falls past the cut (#185 carried forward). The
 * caller arranges first (pinned → non-dismissed → dismissed) and filters before
 * this, so batching applies to exactly what the user is looking at (#667).
 *
 * `hasMore` is true when arranged rows remain beyond what's shown, driving the
 * "Show more" affordance. A pure view-state helper.
 */
export function selectSessionOverflow(
  arranged: Session[],
  activeSessionId: string | null,
  revealCount: number,
): { visible: Session[]; hasMore: boolean } {
  if (revealCount >= arranged.length) {
    return { visible: arranged, hasMore: false };
  }
  const head = arranged.slice(0, revealCount);
  // Pull the active session in even when it sits past the reveal cut, so the
  // current session is never hidden behind "Show more".
  if (activeSessionId && !head.some((s) => s.id === activeSessionId)) {
    const active = arranged.find((s) => s.id === activeSessionId);
    if (active) head.push(active);
  }
  // `hasMore` only when rows still remain hidden after the pulled-in active — so
  // pulling in the single overflow row (active) doesn't strand a "Show more" that
  // reveals nothing.
  return { visible: head, hasMore: head.length < arranged.length };
}

File: apps/cli/src/sessions.rs
//! Pure helpers for the CLI session commands (`sessions list`, `fork`, `chat
//! --resume`), mirroring `apps/desktop/src/lib/sessions.ts` so the two surfaces
//! stay at label-resolution and fork-naming parity (#1080). Kept free of IO and
//! clap so the helpers are unit-testable in isolation — the command layer in
//! `main.rs` owns the store + dispatch, these own the rendering text and the
//! `(Fork N)` numbering algorithm.

use ff_core::Session;

/// The label a user sees for a session: persisted title > goal > fallback.
/// Mirrors `resolveLabel` in `apps/desktop/src/lib/sessions.ts` so the CLI's
/// `sessions list` shows the same text the desktop sidebar would.
pub fn resolve_label(session: &Session) -> &str {
    if let Some(title) = &session.title {
        if !title.is_empty() {
            return title;
        }
    }
    if let Some(goal) = &session.goal {
        if !goal.is_empty() {
            return goal;
        }
    }
    "New session"
}

/// Strip a trailing " (Fork <k>)" suffix (#1069) so forking a fork renumbers
/// from the original base title instead of stacking suffixes. Mirrors
/// `stripForkSuffix` in `apps/desktop/src/lib/sessions.ts`. Hand-rolled (no
/// `regex` dependency) to match the codebase's YAGNI stance — the pattern is
/// simple enough for literal string ops.
pub fn strip_fork_suffix(title: &str) -> &str {
    // Find the last " (Fork " and check it's followed by digits + ")" at the end.
    if let Some(pos) = title.rfind(" (Fork ") {
        let after = &title[pos + " (Fork ".len()..];
        if let Some(num) = after.strip_suffix(')') {
            if !num.is_empty() && num.chars().all(|c| c.is_ascii_digit()) {
                return &title[..pos];
            }
        }
    }
    title
}

/// If `title` is exactly `<base> (Fork <N>)`, return `N`; otherwise `None`.
/// Used by [`next_fork_title`] to find the highest existing fork number
/// sharing a base. Literal string matching (no regex escaping needed, unlike
/// the desktop's TS port — `starts_with` + suffix check is exact).
fn match_fork_number(title: &str, base: &str) -> Option<u32> {
    if !title.starts_with(base) {
        return None;
    }
    let suffix = &title[base.len()..];
    let rest = suffix.strip_prefix(" (Fork ")?;
    let num_str = rest.strip_suffix(')')?;
    num_str.parse::<u32>().ok()
}

/// Compute the next "<base> (Fork N)" title for a session forked from
/// `source_title` (#1069). `base` strips any existing "(Fork k)" suffix from
/// the source so forking a fork stays on the same base and keeps numbering
/// contiguous; `N` is one past the highest existing "(Fork N)" sharing that
/// base among `existing_titles` (the in-memory session list), or 1 if none.
/// Mirrors `nextForkTitle` in `apps/desktop/src/lib/sessions.ts`.
pub fn next_fork_title(source_title: &str, existing_titles: &[Option<&str>]) -> String {
    let base = strip_fork_suffix(source_title);
    let mut max = 0u32;
    for t in existing_titles.iter().flatten() {
        if let Some(n) = match_fork_number(t, base) {
            max = max.max(n);
        }
    }
    format!("{base} (Fork {})", max + 1)
}

/// Format an epoch-millis timestamp as a short local-time string for `sessions
/// list`. Falls back to the raw millis if the timestamp is out of chrono's
/// representable range (same resilience as `fmt_ts` in `ff-session/src/lib.rs`).
pub fn format_ts(ms: i64) -> String {
    chrono::DateTime::from_timestamp_millis(ms)
        .map(|dt| dt.format("%Y-%m-%d %H:%M UTC").to_string())
        .unwrap_or_else(|| ms.to_string())
}

/// Render a list of sessions as TSV (id, label, status, updated-at) to `out`,
/// most-recently-updated first (the store's `list_sessions` order). One header
/// row, then one row per session. Mirrors `ff config list`'s machine-parseable
/// convention. Pure over `(sessions, writer)` so the rendering is unit-testable
/// without a real store or stdout capture.
pub fn render_list(sessions: &[Session], out: &mut impl std::io::Write) -> std::io::Result<()> {
    writeln!(out, "id\tlabel\tstatus\tupdated")?;
    for s in sessions {
        writeln!(
            out,
            "{}\t{}\t{}\t{}",
            s.id,
            resolve_label(s),
            status_label(s.status),
            format_ts(s.updated_at),
        )?;
    }
    Ok(())
}

/// Map a `SessionStatus` to the short lowercase label used in `sessions list`,
/// mirroring the serde `rename_all = "lowercase"` on the enum so the wire form
/// and the CLI text agree.
fn status_label(status: ff_core::SessionStatus) -> &'static str {
    match status {
        ff_core::SessionStatus::Active => "active",
        ff_core::SessionStatus::Done => "done",
        ff_core::SessionStatus::Abandoned => "abandoned",
    }
}

#[cfg(test)]
mod tests;
